feat: Forms store state in redux

vfried opened the pull request:
I created a new model in redux called form draft, this stores the draft for each form.

The forms are stored in redux and changes are done there instead of in the local state as before.

All forms built with the form generator were changed.

Add dataset from project to project
Import dataset
Create dataset
Edit dataset
Create Issue

The biggest change is in the form generator since this change allows the user to leave the form(i.e Create dataset), perform other operations and then go back to the form and continue working. While the user leaves the form there are events that change the state of the form that is currently not mounted.

Further work:
- Improve the way error messages are set/display. There is presentational code inside the component code (in the callbacks). Maybe we should set error codes or something similar so we don't have that code inside the component.
- Refactor the FileUploader so that it only has 1 list of files, it was harder to do with hooks but it shouldn't be that hard now.

Closes #1216 

/deploy

vfried commented on client/src/utils/formgenerator/FormGenerator.state.js:
```diff
@@ -0,0 +1,239 @@
+    return true;
+  }
+
+  setSubmitLoader(location, submitLoader) {
```
The submit loader is the loader that appears at the bottom of the form when you press submit.... It's an object because it basically says weather is active and what text it contains.

![image](https://user-images.githubusercontent.com/42647877/106914051-ed146900-6704-11eb-8c11-4dc4b0f59c63.png)

ciyer replied:
Ah, I see! Could we call it `formStatus`? I think that would be clearer.

vfried replied:
I'm not sure about this, the text and loader activate when the form is submitted. That's why I chose this name. For me, form status could also be related to errors, etc....

vfried commented on client/src/utils/formgenerator/FormGenerator.state.js:
```diff
@@ -0,0 +1,239 @@
+   * @param {string} currentFormModel - current state of the form model
+   */
+  addFormDraft(location, currentFormModel, mounted) {
+    const drafts = this.model.get("formDrafts");
```
This is a good point, I had the idea of doing that but got caught in other things. I will change it to be a dictionary or object.

vfried replied:
Fixed

ciyer replied:
The code is much easier to understand this way!

vfried commented on client/src/utils/formgenerator/FormGenerator.state.js:
```diff
@@ -0,0 +1,239 @@
+
+  setSecondaryButtonText(location, text) {
+    const formDrafts = this.model.get("formDrafts");
+    const draftIndex = formDrafts ? formDrafts.findIndex(elem=> elem.location === location) : -1;
```
Fixed

vfried commented on client/src/project/datasets/change/DatasetChange.container.js:
```diff
@@ -38,39 +39,103 @@ function ChangeDataset(props) {
   };
 
+  const getServerWarnings = (jobsStats) => {
+    const failed = jobsStats.failed
```
This code is now in the presentation component, depending on the form it's possible to have a function to format errors and warnings in a custom way, this allows us to add links, etc...  if this function isn't sent then the errors and warnings are formated in a generic way.

vfried commented on client/src/utils/formgenerator/FormGenerator.container.js:
```diff
@@ -0,0 +1,149 @@
+
+  constructor(props) {
+    super(props);
+    this.model = props.model_top.subModel("formGenerator");
```
Done

vfried commented on client/src/utils/formgenerator/FormGenerator.state.js:
```diff
@@ -0,0 +1,239 @@
+    const formDrafts = this.model.get("formDrafts");
+    const draftIndex = formDrafts ? formDrafts.findIndex(elem=> elem.location === location) : -1;
+    const currentFormDraft = draftIndex > -1 ? formDrafts[draftIndex] : undefined;
+    if (currentFormDraft) {
```
This doesn't apply since the code changed.

ciyer commented on client/src/utils/formgenerator/FormGenerator.state.js:
```diff
@@ -0,0 +1,175 @@
+  removeFormDraft(location) {
+    const drafts = this.model.get("formDrafts");
+    delete drafts[location];
+    let updateObject = { formDrafts: { $set: { ...drafts } } };
```
Is it not possible to do something like?:
```
this.model.set(`formDrafts.{location}`, null)
```

ciyer commented on client/src/utils/formgenerator/FormGenerator.state.js:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  setDraftProperty(location, propertyName, value) {
+    const drafts = this.model.get("formDrafts");
```
Here, I think you can also do something like
```
this.model.set(`formDrafts.{location}.{propertyName}`, value)
```

ciyer reviewed (dismissed):
I like the switch to using maps -- the code is easier to understand. I think it would be possible to work with the model in a simpler way, though.

lorenzo-cavazzi commented on client/src/utils/formgenerator/FormGenerator.container.js:
```diff
@@ -0,0 +1,150 @@
+    this.coordinator = new FormGeneratorCoordinator(props.client, this.model);
+    this.handlers = {
+      addDraft: this.addDraft.bind(this),
+      getDraft: this.getDraft.bind(this),
```
I noticed you created a few `get` functions to get data for the current form, but you also map the same information in `mapStateToProps` so that it's accessible from the properties.
For example, in the `FormPanel`, the content you get from `handlers.getDraft(formLocation)` is the same you have in `drafts[formLocation].currentFormModel`.
I would remove the unneeded functions and access the properties directly.

lorenzo-cavazzi commented on client/src/utils/formgenerator/FormGenerator.container.js:
```diff
@@ -0,0 +1,150 @@
+  mapStateToProps(state) {
+    return {
+      handlers: this.handlers,
+      drafts: state.formGenerator,
```
The type of object stored in the redux store seems to change. `drafts` property starts as an empty array, and then it's overwritten by an object.
I would avoid changing the type to make the code more robust (e.g. if it's always an object, I can use `Object.keys()` safely, but if it's an array that would make the UI crash, and I would need to always test the object type first).

vfried replied:
Fixed

lorenzo-cavazzi commented on client/src/utils/formgenerator/FormGenerator.container.js:
```diff
@@ -0,0 +1,150 @@
+  }
+
+  mapStateToProps(state) {
+    return {
```
Every `FormGenerator` instance seems to be concerned only about his own data, so the `formLocation` is fixed. I wouldn't map all the drafts but only the current one. That should simplify accessing the form models because you won't need to provide the `formLocation` every time.
```
mapStateToProps(state) {
  return {
    draft: state.formGenerator.formDrafts[this.props.formLocation]
  };
}
```
P.S. currently, the `state.formGenerator.formDrafts` starts as an array, so the above code would crash. You can prevent that with `draft: !Array.isArray(state.formGenerator.formDrafts) ? state.formGenerator.formDrafts[this.props.formLocation] : null`

lorenzo-cavazzi commented on client/src/utils/formgenerator/FormGenerator.state.js:
```diff
@@ -0,0 +1,175 @@
+
+class FormGeneratorCoordinator {
+
+  constructor(client, model) {
```
All the functions here accept a `location` parameter, and that seems to be fixed for every instance of `FormGeneratorCoordinator`.
We could move the `location` to the constructor, save it as `this.location` and use that wherever needed to access `formDrafts[location]`.

lorenzo-cavazzi reviewed (changes requested):
It's great to see this feature in action!

There are a couple of potential simplifications in the `FormGenerator` that would simplify the code. I wrote more details inline, but the general idea is:

1. We don't need a `getXxx()` function if we already map that value to a property in `mapStateToProps()`.
2. Every form has its own fixed location and it needs to access only its own data. We could map directly `formDrafts[formLocation]` instead of mapping all the drafts and accessing them per key. We could also move the `location` parameter from the coordinator's functions to its constructor.

vfried commented:
@lorenzo-cavazzi @ciyer Thanks for your suggestions, I pushed recently the new changes.

I took into account both of your inputs. 
Let me know if you would like me to make further changes.

The latest changes are in a separate commit so it's easier to review.

lorenzo-cavazzi commented:
Testing the latest changes, the UI sometimes crashes when clicking on the "Add Dataset" button (not always). The error seems to be in FormGenerator.state.js@66
I'm wondering if using a hash of the path instead of the path itself would be a safer choice to store the forms information -- not sure if this is the cause of the error, but there may be special chars in the URL we may not want to use as an object key.

![Screenshot_20210210_161847](https://user-images.githubusercontent.com/43481553/107531801-79b79f00-6bbd-11eb-8eb4-5a1c5e52eb1e.png)

vfried commented:
This is a good idea, I think the location is also a bit long sometimes and it makes it hard to debug things in redux.

I just did the change but i will wait for the tests to run before pushing.

lorenzo-cavazzi commented on client/src/utils/formgenerator/FormPanel.js:
```diff
@@ -67,10 +68,39 @@ function SubmitButtonGroup(props) {
+        handlers.addDraft(modelValues);
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
There are some missing dependencies, I fear adding this line to ignore them could lead to unexpected behaviors.
I would either remove the empty array dependency or move this part to the parent container. I think the second is cleaner because you could initialize the empty draft there, so you don't need to pass the `model` to this presentational component.
P.S. to avoid using the deep clone from lodash to get an empty object for the  redux store, I've used the `createInitialized` function on the schema -- here is an example (but feel free to keep the current solution if you prefer that)
 https://github.com/SwissDataScienceCenter/renku-ui/blob/bb78438e4cc76122271ffe575dea1633399b4f5d/client/src/project/Project.state.js#L403

lorenzo-cavazzi commented on .github/workflows/acceptance-tests.yml:
```diff
@@ -75,7 +75,7 @@ jobs:
         TEST_ARTIFACTS_PATH: "tests-artifacts-${{ github.sha }}"
         renku_ui: "@${{ github.head_ref }}"
-        # renku: "@master"
+        renku: "@ui1228-fix-acceptance"
```
If the acceptance tests pass, remember to remove this before merging the PR (the test will fail but it's not preventing the merge -- we will need a new release just after)

lorenzo-cavazzi reviewed (commented):
It's working well for me now :tada: 

I have a final concern on disabling a warning for missing dependencies in a `useEffect` hook instead of fixing it. Moving that part from the presentational component to the parent container may simplify adjusting that -- see the comment inline.
P.S. I am not really an expert on the hooks yet. If you think not listing the dependencies is safe, it's fine.

vfried commented:
@lorenzo-cavazzi I just read your review, I renamed the FormPanel file to FormGenerator.present and I removed the useEffect code from there to be executed in the container component in componentDidMount, since this was the intention i think the code looks cleaner now. If i moved to many things let me know.

lorenzo-cavazzi commented:
Since the acceptance tests passed when targeting the `ui1228-fix-acceptance` branch on Renku (https://github.com/SwissDataScienceCenter/renku-ui/runs/1888243042), I believe we can move on with this PR.
Would you mind changing back the line 78 of ".github/workflows/acceptance-tests.yml" `renku: "@ui1228-fix-acceptance"` before merging this PR?
After that, we should make a UI release and merge the PR in the Renku repository to update the acceptance tests.

lorenzo-cavazzi reviewed (approved):
Great to see this feature in action :rocket:
